src/output.ts

```diff
@@ -0,0 +1,96 @@
+import * as path from 'path';
+import { mkdir, stat, writeFile } from 'mz/fs';
+import * as eventToPromise from 'event-to-promise';
+import { spawn as origSpawn, SpawnOptions } from 'child_process';
+import { GeneratedCode, Role } from './types';
+
+const tsconfig = {
+  version: '2.4.2',
+  compilerOptions: {
+    lib: ['es2017', 'esnext'],
+    target: 'es2017',
+    module: 'commonjs',
+    moduleResolution: 'node',
+    emitDecoratorMetadata: true,
+    checkJs: false,
+    allowJs: false,
+    experimentalDecorators: true,
+    downlevelIteration: true,
+    sourceMap: true,
+    declaration: true,
+    strict: true,
+    outDir: '.',
+  },
+  include: [
+    'src/*.ts',
+  ],
+  exclude: [
+    'node_modules',
+  ],
+};
+
+const spawn = (command: string, args?: ReadonlyArray<string>, options?: SpawnOptions) =>
+  eventToPromise(origSpawn(command, args, options), 'exit');
+
+export class TSOutput {
+  protected npm: (...args: string[]) => Promise<number>;
+  protected tsc: (...args: string[]) => Promise<number>;
+
+  protected constructor(protected readonly genPath: string) {
+    this.npm = (...args: string[]) => spawn('npm', args, { cwd: genPath, stdio: 'inherit' });
+    this.tsc = (...args: string[]) => spawn('tsc', args, { cwd: this.genPath, stdio: 'inherit' });
+  }
+
+  public static async create(genPath) {
+    const st = await stat(genPath);
+    if (!st.isDirectory()) {
+      throw new Error(`output dir: ${genPath} is not a directory`);
+    }
+    return new this(genPath);
+  }
+
+  public async write(
+    name: string,
+    version: string,
+    { code, pkg: basePackage }: GeneratedCode,
+    role: Role
+  ) {
+    try {
+      await mkdir(path.join(this.genPath, 'src'));
```

Comment:
This might have been simpler by using `mkdirp` package.

Reply:
Yeah, but this is acceptable IMO